### PR TITLE
Quick fix for unmapped target properties

### PR DESCRIPTION
**Pull request by agudian**
This PR adds some basic stuff for creating quick fixes and also adds the first quick fix:

For the error/warning message "Unmapped target properties", the quick-fix "Ignore unmapped target properties" adds `@Mapping(target = "<prop>", ignore = true)` for each of the unmapped properties.

The quick fix can handle the following cases:
- if there is already an `@Mappings` annotation present, the new `@Mapping` annotations will be added there
- if mapstruct-jdk8 with the repeatable `@Mapping` is used, than the new annotations are placed directly on the method
- if there is already a single (non-repeatable) `@Mapping` annotation present, it will be wrapped into a new `@Mappings` annotation together with the new annotations
- if no annotation is there, yet, and only one annotation needs to be added, then that `@Mapping` is added directly to the method

I used an early version of this some weeks ago while working on a large number of mappers that I switched to `unmappedTargetPolicy=ERROR` and it was really fun... :smile:

### CODE REVIEW

**Comment by gunnarmorling** [comment]
@agudian Finally looking into this. Another possible fix would be to set the "unmapped target policy" in the `@Mapper` annotation to `IGNORE`. WDYT?

**Comment by gunnarmorling** [comment]
The indentation / line breaks might be improved a bit further.

E.g. that's what I have:

```
@Mappings({
})
```

And that's how it's quick fixed:

```
@Mappings({@Mapping(target = "color", ignore = true)
})
```

Or that's what I have:

```
@Mappings({
    @Mapping(target = "age", ignore = true),
    @Mapping(target = "name", ignore = true)
})
```

And that's how it's quick fixed:

```
@Mappings({
    @Mapping(target = "age", ignore = true),
    @Mapping(target = "name", ignore = true), @Mapping(target = "color", ignore = true)
})
```

File: org.mapstruct.eclipse/src/org/mapstruct/eclipse/internal/quickfix/MapStructQuickFix.java [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+                        compilationUnit.becomeWorkingCopy( null );
+                        compilationUnit.commitWorkingCopy( true, null );
+                        compilationUnit.discardWorkingCopy();
+                        marker.delete();
+                    }
+                }
+            }
+        }
+        catch ( CoreException e ) {
+            e.printStackTrace();
```
Comment (gunnarmorling) [inline org.mapstruct.eclipse/src/org/mapstruct/eclipse/internal/quickfix/MapStructQuickFix.java]: Should this rather be re-thrown?

File: org.mapstruct.eclipse/src/org/mapstruct/eclipse/internal/quickfix/fixes/AddIgnoreTargetMappingAnnotationQuickFix.java [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+ */
+public class AddIgnoreTargetMappingAnnotationQuickFix extends MapStructQuickFix {
+
+    public AddIgnoreTargetMappingAnnotationQuickFix() {
+        super( "Ignore unmapped target properties" );
+    }
+
+    @Override
+    public boolean canFix(IMarker marker) {
+        return getMessage( marker ).startsWith( "Unmapped target propert" );
```
Comment (gunnarmorling) [inline org.mapstruct.eclipse/src/org/mapstruct/eclipse/internal/quickfix/fixes/AddIgnoreTargetMappingAnnotationQuickFix.java]: "propert _y_"

File: org.mapstruct.eclipse/src/org/mapstruct/eclipse/internal/quickfix/MapStructMarkerResolutionGenerator.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+                return findApplicableFixes( mk );
+            }
+        }
+        catch ( CoreException e ) {
+            return noResolutions();
+        }
+        return noResolutions();
+    }
+
+    private static IMarkerResolution[] noResolutions() {
```
Comment (gunnarmorling) [inline org.mapstruct.eclipse/src/org/mapstruct/eclipse/internal/quickfix/MapStructMarkerResolutionGenerator.java]: Could be a constant I guess.

File: org.mapstruct.eclipse/src/org/mapstruct/eclipse/internal/quickfix/fixes/AddIgnoreTargetMappingAnnotationQuickFix.java [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+        // if there is already an @Mappings annotation, add the new @Mapping's there
+        Annotation mappingsAnnotation = findAnnotation( method, MAPPINGS_FQ_NAME );
+        if ( mappingsAnnotation != null ) {
+            return rewrite.getListRewrite(
+                ( (SingleMemberAnnotation) mappingsAnnotation ).getValue(),
+                ArrayInitializer.EXPRESSIONS_PROPERTY );
+        }
+
+        // if repeatable @Mapping's are supported, then add the annotations directly to the method
+        if ( supportsRepeatableMapping( unit ) ) {
```
Comment (gunnarmorling) [inline org.mapstruct.eclipse/src/org/mapstruct/eclipse/internal/quickfix/fixes/AddIgnoreTargetMappingAnnotationQuickFix.java:107]: Should this info be cached?

File: org.mapstruct.eclipse/src/org/mapstruct/eclipse/internal/quickfix/MapStructMarkerResolutionGenerator.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        }
+        return result.toArray( new IMarkerResolution[result.size()] );
+    }
+
+    private static Collection<? extends MapStructQuickFix> allQuickFixes() {
+        return Arrays.asList(
+                     new AddIgnoreTargetMappingAnnotationQuickFix()
+                     );
+    }
+}
```
Comment (gunnarmorling) [inline org.mapstruct.eclipse/src/org/mapstruct/eclipse/internal/quickfix/MapStructMarkerResolutionGenerator.java]: Not overly important, but a new line is missing at EOF (same for some other files).

**Comment by gunnarmorling** [comment]
@agudian, great stuff, works like a charm! I've added some comments, but nothing earth-shaking. Looks good overall!

File: org.mapstruct.eclipse/src/org/mapstruct/eclipse/internal/quickfix/MapStructMarkerResolutionGenerator.java (hunk 3 above)
Comment (agudian) [inline org.mapstruct.eclipse/src/org/mapstruct/eclipse/internal/quickfix/MapStructMarkerResolutionGenerator.java]: Right :+1:

File: org.mapstruct.eclipse/src/org/mapstruct/eclipse/internal/quickfix/MapStructMarkerResolutionGenerator.java (hunk 5 above)
Comment (agudian) [inline org.mapstruct.eclipse/src/org/mapstruct/eclipse/internal/quickfix/MapStructMarkerResolutionGenerator.java]: I think we are missing the checkstyle config anyway. I'll add that later on and fix those style issues then.

File: org.mapstruct.eclipse/src/org/mapstruct/eclipse/internal/quickfix/MapStructQuickFix.java (hunk 1 above)
Comment (agudian) [inline org.mapstruct.eclipse/src/org/mapstruct/eclipse/internal/quickfix/MapStructQuickFix.java]: I'm not sure. I'll try out what happens in the UI in that case.

File: org.mapstruct.eclipse/src/org/mapstruct/eclipse/internal/quickfix/fixes/AddIgnoreTargetMappingAnnotationQuickFix.java (hunk 2 above)
Comment (agudian) [inline org.mapstruct.eclipse/src/org/mapstruct/eclipse/internal/quickfix/fixes/AddIgnoreTargetMappingAnnotationQuickFix.java]: That's the lazy way to match on both property and properties... :wink:

File: org.mapstruct.eclipse/src/org/mapstruct/eclipse/internal/quickfix/fixes/AddIgnoreTargetMappingAnnotationQuickFix.java (hunk 4 above)
Comment (agudian) [inline org.mapstruct.eclipse/src/org/mapstruct/eclipse/internal/quickfix/fixes/AddIgnoreTargetMappingAnnotationQuickFix.java:107]: Not sure what the lifecycle of those classes/classloaders here is. And it needs to work even if there are projects with and without Java8 in the workspace. Let's leave this as an optimization for later ;-)

**Comment by agudian** [comment]
> Another possible fix would be to set the "unmapped target policy" in the @Mapper annotation to IGNORE. WDYT?

Good idea. Let's track that in a separate issue.

On your formatting remarks: as you can see in the code, I only add nodes to the AST, not the source text itself. The source should be rendered using your specific formatter-settings. Check the line-breaks for annotation members and array-initializers... That should do the trick for you :-)

Thanks for the review! I'll work in some of the suggestions, add a commit for adding checkstyle to the build and push everything to master then... Some time tonight ;-)

File: org.mapstruct.eclipse/src/org/mapstruct/eclipse/internal/quickfix/MapStructQuickFix.java (hunk 1 above)
Comment (agudian) [inline org.mapstruct.eclipse/src/org/mapstruct/eclipse/internal/quickfix/MapStructQuickFix.java]: Tried it out: the user isn't bothered with the error - it just finds its way into the error-log. So I'll rethrow it here.

**Pull request merged**